cds-plugin.js

```diff
@@ -3,29 +3,69 @@
 
 const cds = require("@sap/cds");
 const cdsPackage = require("@sap/cds/package.json");
-const { initializeFeatures } = require("./src/singleton");
+const { initializeFeatures, getFeaturesKeys, getFeatureValue } = require("./src/singleton");
 
-const activate = async () => {
-  const envFeatureToggles = cds.env.featureToggles;
-  if (envFeatureToggles?.config || envFeatureToggles?.configFile) {
-    // TODO this is currently done in package.json, because "cds build" ignores it otherwise. However, it should happen
-    //  dynamically.
-    // cds.env.requires["FeatureService"] = { model: "@cap-js-community/feature-toggle-library" };
-
-    if (Array.isArray(envFeatureToggles.serviceAccessRoles)) {
-      cds.on("loaded", (csn) => {
-        if (csn.definitions.FeatureService) {
-          csn.definitions.FeatureService["@requires"] = envFeatureToggles.serviceAccessRoles;
-        }
-      });
+const FEATURE_KEY_REGEX = /\/fts\/([^\s/]+)$/;
+
+const _overwriteServiceAccessRoles = (envFeatureToggles) => {
+  if (!Array.isArray(envFeatureToggles.serviceAccessRoles)) {
+    return;
+  }
+  cds.on("loaded", (csn) => {
+    if (csn.definitions.FeatureService) {
+      csn.definitions.FeatureService["@requires"] = envFeatureToggles.serviceAccessRoles;
     }
+  });
+};
 
-    // TODO for the "cds build" use case, this initialize makes no sense
-    await initializeFeatures({
-      config: envFeatureToggles.config,
-      configFile: envFeatureToggles.configFile,
-    });
+const _registerFeatureProvider = () => {
+  if (!cds.env.requires?.toggles || !cds.middlewares?.before) {
+    return;
+  }
+  const cdsFeatures = getFeaturesKeys().reduce((result, key) => {
+    const match = FEATURE_KEY_REGEX.exec(key);
+    if (match) {
+      const feature = match[1];
+      result.push([key, feature]);
+    }
+    return result;
+  }, []);
+  if (cdsFeatures.length === 0) {
+    return;
+  }
+  const contextAuthIndex = cds.middlewares.before.findIndex((entry) => entry.name === "cds_context_auth");
+  if (contextAuthIndex === -1) {
+    return;
   }
+  cds.middlewares.before.splice(contextAuthIndex + 1, 0, function cds_feature_provider(req, res, next) {
+    const user = cds.context?.user.id;
+    const tenant = cds.context?.tenant;
+    req.features =
+      req.headers.features ||
```

Comment:
merge features from header?

Reply:
They are only on/off features, so I think if you overwrite it with the header, you would want the ability to turn them off as well.

Reply:
yes, maybe..

Reply:
Plus, if you merge, you would have to cover all three formats as well... no, I think this approach is fine for now.